tiledb/sm/query/ast/query_ast.h

```diff
@@ -0,0 +1,276 @@
+  /** The attribute name. */
+  std::string field_name_;
+
+  /** The value data. */
+  ByteVecValue condition_value_data_;
+
+  /** A view of the value data. */
+  UntypedDatumView condition_value_view_;
+
+  /** The comparison operator. */
+  QueryConditionOp op_;
+
+ private:
+  static const ASTNodeTag tag_{ASTNodeTag::VAL};
+};
+
+class ASTNodeExpr : public ASTNode {
+ public:
+  ASTNodeExpr(
+      std::vector<tdb_unique_ptr<ASTNode>>&& nodes,
+      QueryConditionCombinationOp c_op)
+      : combination_op_(c_op) {
+    for (const auto& elem : nodes) {
+      nodes_.push_back(elem->clone());
+    }
+  }
+
+  ~ASTNodeExpr() {
+  }
+
+  /** Copy constructor. */
+  ASTNodeExpr(const ASTNodeExpr& rhs)
+      : combination_op_(rhs.combination_op_) {
+    for (const auto& elem : rhs.nodes_) {
+      nodes_.push_back(elem->clone());
+    }
+  };
+
+  /** Move constructor. */
+  ASTNodeExpr(ASTNodeExpr&& rhs)
+      : nodes_(std::move(rhs.nodes_))
+      , combination_op_(rhs.combination_op_){};
+
+  /** Assignment operator. */
+  ASTNodeExpr& operator=(const ASTNodeExpr& rhs) {
+    if (this != &rhs) {
+      for (const auto& elem : rhs.nodes_) {
+        nodes_.push_back(elem->clone());
+      }
+      combination_op_ = rhs.combination_op_;
+    }
+    return *this;
+  }
+
+  /** Move-assignment operator. */
+  ASTNodeExpr& operator=(ASTNodeExpr&& rhs) {
+    nodes_ = std::move(rhs.nodes_);
+    combination_op_ = rhs.combination_op_;
+    return *this;
+  }
+
+  ASTNodeTag get_tag() {
+    return tag_;
+  }
+
+  std::string to_str() {
+    std::string result_str;
+    result_str = "(";
+    for (size_t i = 0; i < nodes_.size(); i++) {
+      auto ptr = nodes_[i].get();
+      if (ptr != nullptr) {
+        result_str += ptr->to_str();
+        if (i != nodes_.size() - 1) {
+          result_str += " ";
+          result_str += query_condition_combination_op_str(combination_op_);
+          result_str += " ";
+        }
+      }
+    }
+    result_str += ")";
+    return result_str;
+  }
+
+  tdb_unique_ptr<ASTNode> clone() {
+    std::vector<tdb_unique_ptr<ASTNode>> nodes_copy;
+    for (const auto& node : nodes_) {
+      nodes_copy.push_back(node->clone());
+    }
+    return tdb_unique_ptr<ASTNode>(
+        tdb_new(ASTNodeExpr, std::move(nodes_copy), combination_op_));
+  }
+
+  /** The node list **/
+  std::vector<tdb_unique_ptr<ASTNode>> nodes_;
+
+  /** The combination operation **/
+  QueryConditionCombinationOp combination_op_;
+
+ private:
+  static const ASTNodeTag tag_{ASTNodeTag::EXPR};
+};
+
+tdb_unique_ptr<ASTNode> ast_combine(
+    const tdb_unique_ptr<ASTNode>& lhs,
+    const tdb_unique_ptr<ASTNode>& rhs,
+    QueryConditionCombinationOp combination_op);
+
+void ast_get_field_names(
+    std::unordered_set<std::string>& field_name_set,
+    const tdb_unique_ptr<ASTNode>& node);
+
+bool ast_is_previously_supported(const tdb_unique_ptr<ASTNode>& node);
+
+}  // namespace sm
+}  // namespace tiledb
+
+#endif  // TILEDB_QUERY_AST_H
```

Comment:
Some of these functions and members should have more documentation.  I am not sure what they do.

Reply:
Resolved? locally?